### ISSUE TITLE
Lora fine tuning phi-3.5 moe

### ISSUE THREAD

**Issue by Manikanta5112** [microsoft/Phi-3CookBook#183]
Hi,

I recently fine-tuned the phi-3.5-moe-instruct model and phi-3.5-mini-instruct model using PEFT LORA. It seems the Moe model is performing way worse than 3.5 Mini Are there any specific things that need to be in mind during LORA fine-tuning with a mixture of expert models? And also during fine-tuning for Moe the validation loss is showing as No Log

**Comment by sujankarki269** [comment]
Can you show me training result and hyperparameter? Did you fine tune using UNSLOTH?

**Comment by Manikanta5112** [comment]
sorry, there is no unsloth support for phi-3.5-moe-instruct model. However, below are the hyper parameters:

   "base_model": "microsoft/Phi-3.5-MoE-instruct",
    "max_seq_length": 4096,

    "lora_config":
    {
        "rank": 32,
        "alpha": 32,
        "task_type": "CAUSAL_LM",
        "target_modules": ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj",], #"all-linear",
        "rslora": True,
        "bias": "lora_only"
    },

    "quantization_config":
    {
        "load_in_4bit": False,
        "bnb_4bit_quant_type": "nf4",
        "bnb_4bit_use_double_quant": True
    },

    "training_arguments":
    {
        "logging_dir": f"logs/{get_iter_name(__file__)}/",
        "output_dir":f"models/{get_iter_name(__file__)}/",
        "evaluation_strategy": "steps",
        "save_strategy": "steps",
        "logging_strategy": "steps",
        "learning_rate": 2e-4, 
        "weight_decay": 0.1,
        "logging_steps": 5,
        "eval_steps": 10,
        "save_steps": 10,
        "eval_delay": 100,
        "warmup_steps": 0,
        "save_total_limit": 5,
        
        "optim": "adamw_torch_fused",
        "per_device_train_batch_size": 2*2*2*2,
        "per_device_eval_batch_size": 2*2*2*2,
        "gradient_accumulation_steps": 4*2*2,
        "eval_accumulation_steps": 4*2*2,
        "gradient_checkpointing": True,

        "adam_beta1": 0.9,
        "adam_beta2": 0.95,
        "adam_epsilon": 1e-8,
        "max_grad_norm": 1.0,
        "lr_scheduler_type": 'cosine',
        "num_train_epochs": 1,
        "continue_from_checkpoint": True,

        "fp16": False,
        "fp16_full_eval": False,

        "bf16": True,
        "bf16_full_eval": True,
    },

**Comment by sujankarki269** [comment]
phi-3.5-mini-instruct model have o_proj and qkv_proj so why are you adding ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj",] in targeted modules?  see the layers in below image and see what are you targetting
![image-6](https://github.com/user-attachments/assets/e031d0bf-a871-40d7-b35f-8a04b56979e8)

**Comment by Manikanta5112** [comment]
I am not using phi-3.5-mini I am using phi-3.5-moe-instruct model

**Comment by sofyc** [comment]
> phi-3.5-mini-instruct model have o_proj and qkv_proj so why are you adding ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj",] in targeted modules? see the layers in below image and see what are you targetting. It too doesnot have "gate_proj", "up_proj" , it has gate_up_proj.
> 
> ![image-6](https://private-user-images.githubusercontent.com/40241326/366854892-e031d0bf-a871-40d7-b35f-8a04b56979e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxODA4NTksIm5iZiI6MTcyNjE4MDU1OSwicGF0aCI6Ii80MDI0MTMyNi8zNjY4NTQ4OTItZTAzMWQwYmYtYTg3MS00MGQ3LWIzNWYtOGEwNGI1Njk3OWU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDIyMzU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyNDk3NjZiOTQzMDJjZDQ3MjNhZmZlY2YzMWIzNTgzOGJkYjZkOTY5ZDI1YmM4YTkyNmExYWUzMmQ1ZDNlMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zcxpck4KzTZsG3pMV5i6A3pHo37-tRlnMNzTDr3ztQo)
> 
> and i too fine tuned phi-3.5-mini-instruct model. it is giving validation loss results tooo... ![image-7](https://private-user-images.githubusercontent.com/40241326/366857265-a3dc3129-aed2-4071-8d54-1dbcae6093b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxODA4NTksIm5iZiI6MTcyNjE4MDU1OSwicGF0aCI6Ii80MDI0MTMyNi8zNjY4NTcyNjUtYTNkYzMxMjktYWVkMi00MDcxLThkNTQtMWRiY2FlNjA5M2I3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDIyMzU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhMDNkMmMwYjU3ZDdhYjc0MGZjMDFlN2YxOTU0MWQ2ZDFiYTMzZDkwYWJhYjY1OTI3MDdjYWU3MDZkOGVlMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WsQZ72C9QFFbsnvJ3Ja5g_LO3pYSTc3hVjnOefLyTY4)

https://github.com/microsoft/Phi-3CookBook/blob/main/code/04.Finetuning/Phi-3-finetune-qlora-python.ipynb

The example in the example scripts in this cookbook appears to be misleading, as it shows

`target_modules = ['k_proj', 'q_proj', 'v_proj', 'o_proj', 'gate_proj', 'down_proj', 'up_proj']`,

which is incorrect based on the model's architecture.

**Comment by sujankarki269** [comment]
![image](https://github.com/user-attachments/assets/9a95c24d-74f1-4d04-928d-1f101c6b1d67)
when i fine tune like this only o_proj is adjusted by lora
![image](https://github.com/user-attachments/assets/558a5166-e23f-4cf4-afa1-4fe8cf4aa518)
